### DISCUSSION TITLE
mamba init & new Mac

### DISCUSSION THREAD

**Discussion by joe61vette** [idaholab/moose#24688] [Q&A Getting Started] [answered]
Hello:

I just began the install process on a new Mac with Ventura 13.4.  All works well until the "mamba activate moose" step.  Then I get the 
"Run 'mamba init' to be able to run mamba activate/deactivate
and start a new shell session. Or use conda to activate/deactivate."
message.  Note that this also happens for a "testing" environment as per the troubleshooting instructions.

I have deleted and re-installed mamba a number of time with the same results.  I pasted the following into my .zshrc file:
```
`# >>> conda initialize >>>
# !! Contents within this block are managed by 'conda init' !!
__conda_setup="$('/Users/joe/mambaforge3/bin/conda' 'shell.zsh' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
    eval "$__conda_setup"
else
    if [ -f "/Users/joe/mambaforge3/etc/profile.d/conda.sh" ]; then
        . "/Users/joe/mambaforge3/etc/profile.d/conda.sh"
    else
        export PATH="/Users/joe/mambaforge3/bin:$PATH"
    fi
fi
unset __conda_setup
# <<< conda initialize <<<

mamba activate moose

export HOST=$(hostname)`
```
My .condarc file is
```
`channels: [https://conda.software.inl.gov/public, conda-forge]
```

If I do a 'mamba info', I get:
```
(base) ~ % mamba info

                  __    __    __    __
                 /  \  /  \  /  \  /  \
                /    \/    \/    \/    \
███████████████/  /██/  /██/  /██/  /████████████████████████
              /  / \   / \   / \   / \  \____
             /  /   \_/   \_/   \_/   \    o \__,
            / _/                       \_____/  `
            |/
        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝

        mamba (1.4.1) supported by @QuantStack

        GitHub:  https://github.com/mamba-org/mamba
        Twitter: https://twitter.com/QuantStack

█████████████████████████████████████████████████████████████


     active environment : base
    active env location : /Users/joe/mambaforge3
            shell level : 1
       user config file : /Users/joe/.condarc
 populated config files : /Users/joe/mambaforge3/.condarc
                          /Users/joe/.condarc
          conda version : 23.1.0
    conda-build version : not installed
         python version : 3.10.10.final.0
       virtual packages : __archspec=1=arm64
                          __osx=13.4=0
                          __unix=0=0
       base environment : /Users/joe/mambaforge3  (writable)
      conda av data dir : /Users/joe/mambaforge3/etc/conda
  conda av metadata url : None
           channel URLs : https://conda.software.inl.gov/public/osx-arm64
                          https://conda.software.inl.gov/public/noarch
                          https://conda.anaconda.org/conda-forge/osx-arm64
                          https://conda.anaconda.org/conda-forge/noarch
          package cache : /Users/joe/mambaforge3/pkgs
                          /Users/joe/.conda/pkgs
       envs directories : /Users/joe/mambaforge3/envs
                          /Users/joe/.conda/envs
               platform : osx-arm64
             user-agent : conda/23.1.0 requests/2.28.2 CPython/3.10.10 Darwin/22.5.0 OSX/13.4
                UID:GID : 501:20
             netrc file : None
           offline mode : False

(base) ~ % 
```
Any suggestions?  Is there some way to completely begin anew? Or does this have to do with Mac OS 13.4?

Thanks,
  Joe Kelly

**Comment by GiudGiud** [comment]
Hello
So the issue that it s always asks for init even though you've done `mamba init` closed the terminal and re-opened already

You are working with zsh not bash right?

**Reply by joe61vette** [reply]
Hi Guillaume.  Yes.  I am working on zsh. And it continually keeps asking for init.

**Reply by GiudGiud** [reply]
The mamba activate moose in your .zshrc also asks for an init?
If there is anything derailing or returning early from the .zshrc that could prevent the initialize part from running

@milljm

**Reply by joe61vette** [reply]
I get:

(base) ~ % mamba activate moose
Run 'mamba init' to be able to run mamba activate/deactivate
and start a new shell session. Or use conda to activate/deactivate.

and

(base) ~ % mamba init

                  __    __    __    __
                 /  \  /  \  /  \  /  \
                /    \/    \/    \/    \
███████████████/  /██/  /██/  /██/  /████████████████████████
              /  / \   / \   / \   / \  \____
             /  /   \_/   \_/   \_/   \    o \__,
            / _/                       \_____/  `
            |/
        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝

        mamba (1.4.1) supported by @QuantStack

        GitHub:  https://github.com/mamba-org/mamba
        Twitter: https://twitter.com/QuantStack

█████████████████████████████████████████████████████████████

no change     /Users/joe/mambaforge3/condabin/conda
no change     /Users/joe/mambaforge3/bin/conda
no change     /Users/joe/mambaforge3/bin/conda-env
no change     /Users/joe/mambaforge3/bin/activate
no change     /Users/joe/mambaforge3/bin/deactivate
no change     /Users/joe/mambaforge3/etc/profile.d/conda.sh
no change     /Users/joe/mambaforge3/etc/fish/conf.d/conda.fish
no change     /Users/joe/mambaforge3/shell/condabin/Conda.psm1
no change     /Users/joe/mambaforge3/shell/condabin/conda-hook.ps1
no change     /Users/joe/mambaforge3/lib/python3.10/site-packages/xontrib/conda.xsh
no change     /Users/joe/mambaforge3/etc/profile.d/conda.csh
modified      /Users/joe/.bash_profile

==> For changes to take effect, close and re-open your current shell. <==

Added mamba to /Users/joe/.bash_profile

==> For changes to take effect, close and re-open your current shell. <==

(base) ~ % 

Note that it added mamba to bash.  Does that make a difference?

Then I close terminal window & open another.

Last login: Wed Jun 14 14:02:00 on ttys001
Run 'mamba init' to be able to run mamba activate/deactivate
and start a new shell session. Or use conda to activate/deactivate.

(base) ~ % 

Note there is a 'mamba activate moose' in my .zshrc file.

**Reply by GiudGiud** [reply]
can you paste a screenshot of a fresh new terminal?

mamba thinks you are using bash right now.

**Comment by joe61vette** [comment]
I killed terminal and then relaunched it.  The screenshot is attached.
<img width="789" alt="Screenshot 2023-06-14 at 3 03 36 PM" src="https://github.com/idaholab/moose/assets/8108870/13383e12-af53-4189-bdf8-6e6a69789cb2">

**Reply by joe61vette** [reply]
(base) ~ % echo $0
-zsh

**Reply by milljm** [reply]
I have seen this before. And while everything looks like zsh, the system default is still set to BASH. I like to blame Apple on this, as even their instructions:

```bash
The default interactive shell is now zsh.
To update your account to use zsh, please run `chsh -s /bin/zsh`.
For more details, please visit https://support.apple.com/kb/HT208050.
```
...doesn't work. You have to do it using their GUI. That is, through System Preferences Ref: https://github.com/idaholab/moose/discussions/24072#discussioncomment-6064726

Can you verify that your default profile is Zsh? By way of using System Preferences, Users & Groups, Right-click on you, Advanced Options... and this is where it gets a little fuzzy for us. We no longer even have the ability to get this far, and leave screen shots for what we are trying to demonstrate.

**Reply by joe61vette** [reply]
Here is a screenshot.  It says zsh.
<img width="466" alt="Screenshot 2023-06-14 at 3 31 46 PM" src="https://github.com/idaholab/moose/assets/8108870/ece45e4b-707c-44e5-8a7b-96bf0bf5e3f1">

**Reply by joe61vette** [reply]
I just tried to source a mamba_init file.  Nothing changed.  The file is attached.
[mamba_init.txt](https://github.com/idaholab/moose/files/11749943/mamba_init.txt)

**Reply by milljm** [reply]
Are you comfortable removing the entry Mamba added to your ~/.bash_profile? If so, lets do that... open a new term, see what happens.

I am asking that you remove this entire piece from ~/.bash_profile:

```bash
# >>> conda initialize >>>
# !! Contents within this block are managed by 'conda init' !!
__conda_setup="$('/Users/joe/mambaforge3/bin/conda' 'shell.bash' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
    eval "$__conda_setup"
else
    if [ -f "/Users/joe/mambaforge3/etc/profile.d/conda.sh" ]; then
        . "/Users/joe/mambaforge3/etc/profile.d/conda.sh"
    else
        export PATH="/Users/joe/mambaforge3/bin:$PATH"
    fi
fi
unset __conda_setup
# <<< conda initialize <<<
```

Apologies if you've done this already. I am just arriving to this thread.

**Reply by joe61vette** [reply]
Unfortunately.  No difference.  The steps I did:
1) edit .bash_profile (basically erased everything
2) quit terminal and reopen
3) mamba init
4) open a new terminal window
5) mamba activate moose, this again says to 'mamba init'

BUT, I went back to my bash_profile and the following had been added again:

# >>> conda initialize >>>
# !! Contents within this block are managed by 'conda init' !!
__conda_setup="$('/Users/joe/mambaforge3/bin/conda' 'shell.bash' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
    eval "$__conda_setup"
else
    if [ -f "/Users/joe/mambaforge3/etc/profile.d/conda.sh" ]; then
        . "/Users/joe/mambaforge3/etc/profile.d/conda.sh"
    else
        export PATH="/Users/joe/mambaforge3/bin:$PATH"
    fi
fi
unset __conda_setup

if [ -f "/Users/joe/mambaforge3/etc/profile.d/mamba.sh" ]; then
    . "/Users/joe/mambaforge3/etc/profile.d/mamba.sh"
fi
# <<< conda initialize <<<

**Reply by milljm** [reply]
That seems normal... that happens for me as well:

```bash
(base) ❯      # < ---- I am already initialized in (base)
(base) ❯ echo $0
-zsh
(base) ❯  mamba init

                  __    __    __    __
                 /  \  /  \  /  \  /  \
                /    \/    \/    \/    \
███████████████/  /██/  /██/  /██/  /████████████████████████
              /  / \   / \   / \   / \  \____
             /  /   \_/   \_/   \_/   \    o \__,
            / _/                       \_____/  `
            |/
        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝

        mamba (1.4.1) supported by @QuantStack

        GitHub:  https://github.com/mamba-org/mamba
        Twitter: https://twitter.com/QuantStack

█████████████████████████████████████████████████████████████

no change     /Users/milljm/mambaforge3/condabin/conda
no change     /Users/milljm/mambaforge3/bin/conda
no change     /Users/milljm/mambaforge3/bin/conda-env
no change     /Users/milljm/mambaforge3/bin/activate
no change     /Users/milljm/mambaforge3/bin/deactivate
no change     /Users/milljm/mambaforge3/etc/profile.d/conda.sh
no change     /Users/milljm/mambaforge3/etc/fish/conf.d/conda.fish
no change     /Users/milljm/mambaforge3/shell/condabin/Conda.psm1
no change     /Users/milljm/mambaforge3/shell/condabin/conda-hook.ps1
no change     /Users/milljm/mambaforge3/lib/python3.10/site-packages/xontrib/conda.xsh
no change     /Users/milljm/mambaforge3/etc/profile.d/conda.csh
modified      /Users/milljm/.bash_profile

==> For changes to take effect, close and re-open your current shell. <==

Added mamba to /Users/milljm/.bash_profile

==> For changes to take effect, close and re-open your current shell. <==
```
It added/edit my ~/.bash_profile. Yet I am operating in Zsh.

What I'd like to know what happens, is when you:

- Open a new terminal
- `mamba activate moose`

Does it activate? Or tell you, you must first initialize mamba?

**Reply by joe61vette** [reply]
From a new terminal window:

Last login: Wed Jun 14 15:50:46 on ttys001
Run 'mamba init' to be able to run mamba activate/deactivate
and start a new shell session. Or use conda to activate/deactivate.

(base) ~ % mamba acivate moose
No command 'conda acivate'.
Did you mean 'conda activate'?

CommandNotFoundError: No command 'conda acivate'.
Did you mean 'conda activate'?

(base) ~ % 

Then 'mamba init', open another window, and it says 

Last login: Wed Jun 14 16:03:54 on ttys000
Run 'mamba init' to be able to run mamba activate/deactivate
and start a new shell session. Or use conda to activate/deactivate.

It is an endless cycle.  I can never activate moose (Note the I do have 'mamba activate moose' just after the Conda initialization stuff in my .zshrc file).

**Reply by joe61vette** [reply]
I went back and redid this correctly spelling activate.  Same result.

**Reply by milljm** [reply]
Endless cycle... Sounds like fun... 😞 

This isn't a solution, but can you try this:
```bash
source activate /Users/joe/mambaforge3/envs/moose
```
(assuming `mamba create -n moose moose-dev` has worked!)

**Reply by joe61vette** [reply]
That works! 

Also, I forgot to mention that when I first started this process, I made a mistake and copied/pasted the mamba install command for intel rather than arm.  So, I deleted that version and started over.  I think that I deleted everything.

**Reply by milljm** [reply]
Well, you `conda info --all` looks fine. So I am sorta at a loss as to whats going on with Mamba initialization...

This can reveal some sensitive information, but can you perform and report your running environment after opening a new terminal?
```bash
env
```
Please look through it, and just redact anything you feel is something we shouldn't know.

**Reply by joe61vette** [reply]
Here it is:

(base) ~ % env
__CFBundleIdentifier=com.apple.Terminal
TMPDIR=/var/folders/_d/lcs69b3x4nn1x344gcw141200000gn/T/
XPC_FLAGS=0x0
TERM=xterm-256color
SSH_AUTH_SOCK=/private/tmp/com.apple.launchd.foE50SVnzb/Listeners
XPC_SERVICE_NAME=0
TERM_PROGRAM=Apple_Terminal
TERM_PROGRAM_VERSION=447
TERM_SESSION_ID=F505A9AA-69AB-4686-A1DC-EAD38DE5C5E6
SHELL=/bin/zsh
HOME=/Users/joe
LOGNAME=joe
USER=joe
PATH=/Users/joe/mambaforge3/bin:/Users/joe/mambaforge3/condabin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/TeX/texbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin
SHLVL=1
PWD=/Users/joe
OLDPWD=/Users/joe
PROMPT=%1~ %# 
CLICOLOR=1
LSCOLORS=gxBxhxDxfxhxhxhxhxcxcx
CONDA_EXE=/Users/joe/mambaforge3/bin/conda
_CE_M=
_CE_CONDA=
CONDA_PYTHON_EXE=/Users/joe/mambaforge3/bin/python
CONDA_SHLVL=1
CONDA_PREFIX=/Users/joe/mambaforge3
CONDA_DEFAULT_ENV=base
CONDA_PROMPT_MODIFIER=(base) 
XML_CATALOG_FILES=file:///Users/joe/mambaforge3/etc/xml/catalog file:///etc/xml/catalog
HOST=...I deleted this just in case...
LANG=en_US.UTF-8
_=/usr/bin/env

**Reply by milljm** [reply]
nothing out of the ordinary that I can see.

We can try to add some print statements... See where we land during initialization attempts and go from there:

```bash
# >>> conda initialize >>>
# !! Contents within this block are managed by 'conda init' !!
__conda_setup="$('/Users/joe/mambaforge3/bin/conda' 'shell.zsh' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
    printf "no error in hook\n"
    eval "$__conda_setup"
else
    if [ -f "/Users/joe/mambaforge3/etc/profile.d/conda.sh" ]; then
        printf "sourcing\n"
        . "/Users/joe/mambaforge3/etc/profile.d/conda.sh"
    else
        printf "doing nothing\n"
        export PATH="/Users/joe/mambaforge3/bin:$PATH"
    fi
fi
unset __conda_setup
# <<< conda initialize <<<
```
(those three `printf` statements)

**Reply by joe61vette** [reply]
Here is the output:

Last login: Wed Jun 14 16:27:44 on ttys000
no error in hook
Run 'mamba init' to be able to run mamba activate/deactivate
and start a new shell session. Or use conda to activate/deactivate.

**Reply by milljm** [reply]
I am able to replicate your issue when I do the following:

```bash
(base) env -i zsh -f   # starts a new zsh environment, clean of any profiles
INL618984%
INL618984% junk="$('/Users/milljm/mambaforge3/bin/conda' 'shell.zsh' 'hook')"
INL618984% eval ${junk}
(base) INL618984%
(base) INL618984% mamba activate moose
Run 'mamba init' to be able to run mamba activate/deactivate
and start a new shell session. Or use conda to activate/deactivate.
```
At least I have something I can play around with on my own.

**Reply by milljm** [reply]
Okay, so when mamba is not a function, it calls the binary (well executable script) mamba within your PATH. Which has that message hard-coded:

```python
# Main entry point!
def main(*args, **kwargs):
    # Set to false so we don't allow uploading our issues to conda!
    context.report_errors = False

    from conda.common.compat import ensure_text_type
    if "activate" in sys.argv or "deactivate" in sys.argv:
        print(
            "Run 'mamba init' to be able to run mamba activate/deactivate\n"
            "and start a new shell session. Or use conda to activate/deactivate.\n",
            file=sys.stderr,
        )
        return sys.exit(-1)
```

I suspect your `mamba` is not a function then? Can you check?

```bash
(base) INL618984% type mamba
mamba is /Users/milljm/mambaforge3/bin/mamba
```
as opposed to what it wants:
```bash
❯ type mamba
mamba is a shell function from /Users/milljm/mambaforge3/etc/profile.d/mamba.sh
```
I am confused a bit on how Mamba is initialized as a function (er, or rather how it is _not_ initialized). When I figure that out, I'll post again.

**Reply by milljm** [reply]
Indeed...

```bash
(base) INL618984% source /Users/milljm/mambaforge3/etc/profile.d/mamba.sh
(base) INL618984% type mamba
mamba is a shell function from /Users/milljm/mambaforge3/etc/profile.d/mamba.sh
(base) INL618984% mamba activate moose
(moose) INL618984% 
```

So why is your mamba.sh not being 'sourced' properly when you activate the base environment.

**Reply by milljm** [reply, marked as answer]
OOooo. I just realized I have something very important in my initialization profile that you do not:

```bash
# >>> conda initialize >>>
# !! Contents within this block are managed by 'conda init' !!
__conda_setup="$('/Users/milljm/mambaforge3/bin/conda' 'shell.zsh' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
    eval "$__conda_setup"
else
    if [ -f "/Users/milljm/mambaforge3/etc/profile.d/conda.sh" ]; then
        . "/Users/milljm/mambaforge3/etc/profile.d/conda.sh"
    else
        export PATH="/Users/milljm/mambaforge3/bin:$PATH"
    fi
fi
unset __conda_setup

if [ -f "/Users/milljm/mambaforge3/etc/profile.d/mamba.sh" ]; then    <-- where did
    . "/Users/milljm/mambaforge3/etc/profile.d/mamba.sh"              <-- this come
fi                                                                    <-- from??
# <<< conda initialize <<<
```
I have the all-important `mamba` source. With that said, I am willing to bet you can use `conda activate` instead of `mamba activate`.

I also think I know how this came to be 😄 

When you install Conda or Mamba, it will detect the presence of:
```bash
# >>> conda initialize >>>
anything in between these identifiers 
# <<< conda initialize <<<
```
already in your profile and not _modify_ the profile again. So this means you may have had a non-mamba install at some point?

**Reply by joe61vette** [reply]
I add these 2 lines to my .zshrc and all is well.

if [ -f "/Users/joe/mambaforge3/etc/profile.d/mamba.sh" ]; then    
    . "/Users/joe/mambaforge3/etc/profile.d/mamba.sh"              
fi  

It is likely that I had copied a .zshrc file from another computer before installing mamba.  So the 'condo initialize' stuff got skipped.  Why those lines are necessary on this computer and not the other one... That will remain a mystery to me.

THANKS so much for the help.  It is greatly appreciated !

**Reply by joe61vette** [reply]
This is what I got for mamba type:

moose) ~ % type mamba
mamba is a shell function from /Users/joe/mambaforge3/etc/profile.d/mamba.sh

**Reply by milljm** [reply]
Now that its a function, `mamba activate` should work :)

Cheers! Thanks for reporting the issue. I learned some things.

**Comment by jayant-yadav** [comment]
I wrecked my brain for days on this. Until I stumbled upon [this answer](https://community.anaconda.cloud/t/unable-to-activate-environment-prompted-to-run-conda-init-before-conda-activate-but-it-doesnt-work/68677/3?u=jayantyadav202)
All I had to do was to `mamba init zsh` because it needed shell name. 
I am using Intel MacBook Pro.
